### ISSUE TITLE
2 body elements on a page

### ISSUE THREAD

**Issue by obeyer** [ampproject/amphtml#2305]
![screen shot 2016-02-26 at 9 53 26 am](https://cloud.githubusercontent.com/assets/1790156/13360616/ee8583d0-dc6e-11e5-83e0-a4e1052a315c.png)
when you look at the page with #development=1 at the end of the url, it breaks the body element into 2 bodies, therefore changing the structure of the element. is that a bug? or an intention? id this is an intention, how are we expect to code our css, if you broke the structure of the document? first I wasn't worry about it, because I was using #development=1 just for validation, but now that the AMP rolled out, and I can see how the results come up in google search, when I click on the post, it appears in the carousel, and the entire page is in an iframe, and the html document has 2 body tags!

**Comment by cramforce** [comment]
Dima, do you want to explain why? :)

**Comment by dvoytenko** [comment]
@obeyer in short this is due to the following two Safari bugs:

https://bugs.webkit.org/show_bug.cgi?id=154399
https://bugs.webkit.org/show_bug.cgi?id=149264

The bottom line: scrolling and `position:fixed` are not very friendly to each other in Safari.

In long - you can see https://github.com/ampproject/amphtml/blob/master/src/service/fixed-layer.js class that goes over these points. I will also soon publish an article about issues with scrolling in Safari and the steps we've been taking to deal with them.

I will keep this issue open until I publish the article.

**Comment by obeyer** [comment]
@dvoytenko 
why don't you do overflow: scroll and not auto?
also, you can make scrolling an position:fixed worked together in Safari mobile browser. if you look at the amp page not within your carousel with search result, but just in a normal way (same page from my screenshot http://www.popsugar.com/celebrity/Kate-Middleton-Tennis-Workshop-Scotland-2016-40322850/amp), you can see there's a pure css hamburger menu at the top right, which when clicked, displays a slide-out menu, which has position: fixed, and you can scroll on it just all right. and the top nav bar with a logo has position fixed. 
when you parse the document, you add z-indexes which I much higher than we have. you can do with values 1, 2, 3 just fine, because the parents define the scope of z-indexes for their children. You put z-indexes there like 10099. why so high? why to be in a constant race to the maximum value, and then what? Anyways, I can change z-index, and my hamburger menu icon will show even inside the carousel with amp-result, and it will scroll on a position fixed. What doesn't work in that case, is that the nav doesn't move to the right, because you put it in a different body tag, and I can't target it anymore on :hover, because it is not a next closest sibling, and not a child.
Long story short, your problem with scrolling still doesn't explain why you needed to make 2 body tags, which is, first, illegal, and second, changes the markup of the page in an unexpected way, so the css stops working

**Comment by obeyer** [comment]
we have 4 fixed element on a page, and 2 of them got into the first body, and the other 2 are in the second body

**Comment by dvoytenko** [comment]
@obeyer Hmm. Interesting. You think this is specifically related to `overflow:auto` and `scroll` would work better? Let me try it out - everything is possible here.

RE: different elements in different layers. Generally, we can arrange so that, essentially, you can disable this feature for your pages, but we need to tread carefully - the scrolling/position:fixed is often really bad for embedded iOS Safari.

**Comment by dvoytenko** [comment]
@obeyer Well, unfortunately, `overflow:scroll` does work any better than `overflow:auto`. I created a little page that shows what's going on with a web page iframed on iOS: http://jsbin.com/gugika/quiet

You can try it out yourself on iPhone or you can see it in this video: https://drive.google.com/file/d/0ByWTnslqTJj6R21JbTk2NFV2RzQ/view?usp=sharing

You will see how `position:fixed` elements jump around when scrolled. Obviously the experience is very bad and this is the reason for our workaround. I pretty much hate everything about it, but I see no other way at this time. This was the last thing we wanted to do and, literally, was the last thing we had to do.

So what can we do?

1. You can chime in on https://bugs.webkit.org/show_bug.cgi?id=154399
2. In your case, this feature is only really relevant to the bottom (share) bar and I think it works well for it. Otherwise, you could try to switch to `position:absolute` for both the button and sliding panel - I know a few publishers that do that and it works ok.
3. We are moving to implementing side menus as a 1st party component in AMP - you can collaborate with us to create the menu that will work well for you.

**Comment by obeyer** [comment]
@dvoytenko 
look, I might have found the workaround for my menu problem, but I need to be sure that you don't keep changing the way it works. Still need to test it in prod, though...
Also, it would be nice, if the page looked the same with and without #development=1, at least in terms of markup and inline styles that are been added.
Thank you!

**Comment by dvoytenko** [comment]
@obeyer This is a highly frustrated bug and very poorly documented area - so it's hard to promise that we won't change it, but you'd share your approach with us - I can comment on how likely this to break in the future and I can do more testing on our end. The ultimate solution is to support menus as a 1st party component.

We actually now force this "weird" mode via `#development=1` on purpose. Otherwise no one would ever notice this problem until it'd hit PROD.

**Comment by obeyer** [comment]
@dvoytenko 
as per your comment: We actually now force this "weird" mode via #development=1 on purpose. Otherwise no one would ever notice this problem until it'd hit PROD.
What I meant, is that could it be the default mode, even without #development=1? since it is how it looks on prod. please?
my approach to fix my menu was to drop some parts of design elements we have on normal site, like different logo inside the navigation panel. so I dropped that, since I can't move the one on the top nav anymore, because it is out of bounds :) And I had to change z-indexes, and the dark overlay behaviour as well, comparing to a normal mobile theme.
As of an idea to support menus as a 1st party component:  I am not sure about that. Everyone wants their amp site to match their normal mobile design and user experience. And as you can imagine, there are as many designs, as designers out there. If our site design didn't have the fixed header at the top, the navigation implementation would have been much easier. But in any case it is something that can be implemented with pure css.

**Comment by dvoytenko** [comment]
@obeyer 

> What I meant, is that could it be the default mode, even without #development=1? since it is how it looks on prod. please?

Yes, let us think through this. But we would certainly consider this.

> As of an idea to support menus as a 1st party component: I am not sure about that.

Ideally this component will only take care of invoking the menu, setting some CSS classes and take care of some of the `position:fixed` nuances. And everything else would be left to publishers.

**Comment by adelinamart** [comment]
Hey,

The AMP community has been working nonstop to make AMP better, but somehow we've still managed to grow an enormous backlog of open issues.  This has made it difficult for the community to prioritize what we should work on next.

A new process is on the way and to give it a chance for success we will be closing issues that have not been updated in awhile.

If this issue still requires further attention, simply reopen it.  Please verify first whether the feature proposal is still relevant.

We really appreciate the contribution! Thank you for bearing with us as we drag ourselves out of the issue abyss. :)
